src/main/java/net/neoforged/neoforge/network/configuration/CheckFeatureFlags.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.network.configuration;
+
+import com.google.common.collect.Sets;
+import com.mojang.logging.LogUtils;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.configuration.ClientConfigurationPacketListener;
+import net.minecraft.network.protocol.configuration.ServerConfigurationPacketListener;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.network.ConfigurationTask;
+import net.minecraft.world.flag.FeatureFlags;
+import net.neoforged.neoforge.network.handling.IPayloadContext;
+import net.neoforged.neoforge.network.payload.FeatureFlagAcknowledgePayload;
+import net.neoforged.neoforge.network.payload.FeatureFlagDataPayload;
+import org.jetbrains.annotations.ApiStatus;
+import org.slf4j.Logger;
+
+@ApiStatus.Internal
+public record CheckFeatureFlags(ServerConfigurationPacketListener listener) implements ConfigurationTask {
+    public static final Type TYPE = new Type(ResourceLocation.fromNamespaceAndPath("neoforge", "check_feature_flags"));
+    private static final Logger LOGGER = LogUtils.getLogger();
+    private static Set<ResourceLocation> moddedFlags = null;
+
+    @Override
+    public void start(Consumer<Packet<?>> packetSender) {
+        if (listener.getConnection().isMemoryConnection()) {
+            listener.finishCurrentTask(TYPE);
+            return;
+        }
+
+        Set<ResourceLocation> moddedFlags = getModdedFeatureFlags();
+        if (listener.getConnectionType().isOther() || !listener.hasChannel(FeatureFlagDataPayload.TYPE)) {
+            if (!moddedFlags.isEmpty()) {
+                // Use plain components as vanilla connections will be missing our translation keys
+                listener.disconnect(Component.literal("This server does not support vanilla clients as it has custom FeatureFlags"));
+            } else {
+                listener.finishCurrentTask(TYPE);
+            }
+            return;
+        }
+        packetSender.accept(new FeatureFlagDataPayload(moddedFlags).toVanillaClientbound());
+    }
+
+    public static void handleClientboundPayload(FeatureFlagDataPayload payload, IPayloadContext context) {
+        Set<ResourceLocation> localFlags = getModdedFeatureFlags();
+        Set<ResourceLocation> remoteFlags = payload.moddedFlags();
+        if (localFlags.equals(remoteFlags)) {
+            context.reply(FeatureFlagAcknowledgePayload.INSTANCE);
+        } else {
+            context.disconnect(Component.translatable("neoforge.network.extensible_enums.enum_entry_mismatch"));
```

Comment:
Shouldn't this be the feature flag entry mismatch translation and not the enum extension one?
`neoforge.network.feature_flags.entry_mismatch`

Reply:
Nice catch, it definitely should not be the enum one